win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1167,214 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
     return TRUE;    /* success */
 }
 
+static inline VOID
+FontLink_Chain_Init(
+    _Out_ PFONTLINK_CHAIN pChain,
+    _Inout_ PTEXTOBJ pTextObj,
+    _In_ FT_Face face,
+    _In_ PDC dc)
+{
+    RtlZeroMemory(pChain, sizeof(*pChain));
+    pChain->pBaseTextObj = pTextObj;
+    pChain->pDefFace = face;
+    pChain->dc = dc;
+}
+
+static NTSTATUS
+FontLink_Chain_LoadReg(
+    _Inout_ PFONTLINK_CHAIN pChain,
+    _Inout_ LPLOGFONTW pLF)
+{
+    NTSTATUS Status;
+    HKEY hKey;
+    DWORD cbData;
+    static const WCHAR s_szzDefFontLink[] =
+        L"tahoma.ttf,Tahoma\0"
+        L"msgothic.ttc,MS UI Gothic\0"
+        L"mingliu.ttc,PMingLiU\0"
+        L"simsun.ttc,SimSun\0"
+        L"gulim.ttc,Gulim\0"
+        L"\0";
+    static const WCHAR s_szzDefFixedPitchFontLink[] =
+        L"cour.ttf,Courier New\0"
+        L"msgothic.ttc,MS Gothic\0"
+        L"mingliu.ttc,MingLiU\0"
+        L"simsun.ttc,NSimSun\0"
+        L"gulim.ttc,GulimChe\0"
+        L"\0";
+    WCHAR szzFontLink[512]; // 512 is enough size
+    SIZE_T FontLinkSize;
+    PZZWSTR pszzFontLink;
+
+    // Open registry key
+    Status = RegOpenKey(
+        L"\\Registry\\Machine\\Software\\Microsoft\\Windows NT\\CurrentVersion\\FontLink\\SystemLink",
+        &hKey);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    // Load the FontLink entry
+    cbData = sizeof(szzFontLink);
+    Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, szzFontLink, &cbData);
+    if (!NT_SUCCESS(Status) && SubstituteFontRecurse(pLF)) // Retry with substituted name
+    {
+        cbData = sizeof(szzFontLink);
+        Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, szzFontLink, &cbData);
+    }
+
+    ZwClose(hKey); // Close the registry key
+
+    RtlStringCchCopyW(pChain->szFaceName, _countof(pChain->szFaceName), pLF->lfFaceName);
+
+    ASSERT(SZZ_GetSize(s_szzDefFontLink) <= sizeof(szzFontLink));
+    ASSERT(SZZ_GetSize(s_szzDefFixedPitchFontLink) <= sizeof(szzFontLink));
+
+    if (!NT_SUCCESS(Status)) // Registry value not found
+    {
+        if (!(pLF->lfPitchAndFamily & FIXED_PITCH))
+            RtlCopyMemory(szzFontLink, s_szzDefFontLink, SZZ_GetSize(s_szzDefFontLink));
```

Comment:
Since s_szzDef.... is a known fixed constant data array, isn't its size known at compile-time?
If so, `sizeof(s_szzDefFontLink)` should be better.
Same comment for the fixedPitch below too.

Reply:
Fixed.